### ISSUE TITLE
Syntax for using a Maven bom is confusing

### ISSUE THREAD

**Issue by wilkinsona** [gradle/gradle#4422]
### Expected Behavior

The syntax for importing a bom makes it clear that it's not just another dependency and that it will affect your build in different ways to a normal dependency.

When I designed the dependency management plugin, I chose to use a DSL like the following to import a bom into your build:

```
dependencyManagement {
    implementation {
         imports {
              mavenBom 'org.springframework.boot:spring-boot-dependencies:1.5.8.RELEASE'
         }
    }
}
```

I don't claim that this is perfect, but I do believe it's clearer and less confusing than Gradle 4.6 RC1's equivalent which is the following:

```
dependencies {
    // import a BOM
    implementation 'org.springframework.boot:spring-boot-dependencies:1.5.8.RELEASE'
}
```

The `// import a BOM` comment is taken from [the release notes](https://docs.gradle.org/4.6-rc-1/release-notes.html#experimental-bom-import). That it was felt it was needed to clearly explain the new feature is telling. If the DSL were clearer, a comment would not have been necessary.

Another advantage of the dependency management plugin's DSL is that it allows you to import a bom and apply it to all configurations:

```
dependencyManagement {
    implementation {
         imports {
              mavenBom 'org.springframework.boot:spring-boot-dependencies:1.5.8.RELEASE'
         }
    }
}
```

Lastly, using a syntax that's specific for importing boms provides room for extension in the future. One example that I'd like to see if the ability to override properties in the imported bom:

```
dependencyManagement {
    imports {
        mavenBom('org.springframework.boot:spring-boot-dependencies:1.5.8.RELEASE') {
            bomProperty 'spring.version', '4.3.4.RELEASE'
        }
    }
}
```

Again, I'm not saying that this DSL is perfect, but I do think it has some notable benefits over what Gradle 4.6 currently offers.

### Current Behavior

The syntax makes a bom look like any other dependency.

### Context

A bom is unlike a normal dependency. When a bom is used, it can have far-reaching affects on all of a project's dependencies. A bom also shouldn't end up on the classpath like normal dependencies, instead it should influence the versions of dependencies that are on the classpath.

### Steps to Reproduce (for bugs)

N/A

### Your Environment

N/A

**Comment by melix** [comment]
We should leave the properties discussion out of this in the first place. We're very careful not to introduce unnecessary complexity and redundant features, so we think in terms of how to best support _use cases_ rather than how they technically map.

Regarding the "import" syntax, there's no technical need for it. The initial spike we had was using an "import" syntax but we reckoned it had no advantage whatsoever. Especially, it's not clear for everybody what a BOM is, because it's specified out of _what is found in a POM file_ rather than on what it's supposed to represent.

I'd argue that _if_ it's necessary to make the difference, we now have the tooling for this, also shipped in 4.6: for each dependency, you can add a "reason" why it's here:

```
dependencies {
    implementation('org.springframework.boot:spring-boot-dependencies:1.5.8.RELEASE') {
       because "this application relies on the Spring Boot platform"
    }
}
```

The advantage is that this custom reason appears in the dependency insight report. And it helps providing better explanations when you need a more recent release of a library found in a BOM too, which cannot be expressed by just upgrading properties:

```
dependencies {
    implementation('org.springframework.boot:spring-boot-dependencies:1.5.8.RELEASE') {
       because "this application relies on the Spring Boot platform"
    }

    implementation('org.springframework.boot:whatever:1.5.9') {
       because 'release 1.5.8 has a bug which impacts our application'
    }
}
```

In short, we want users to focus on _why_ they need dependencies/BOMs, so that it's easier to understand when something is wrong.

**Comment by melix** [comment]
To be clear, I suggest we change the release notes to do the above.

**Comment by wilkinsona** [comment]
> We should leave the properties discussion out of this in the first place

Properties was purely an example. The broader point was the benefit of having an extensible syntax for dealing with boms.

> Regarding the "import" syntax, there's no technical need for it. The initial spike we had was using an "import" syntax but we reckoned it had no advantage whatsoever.

I think there's a pretty big advantage in terms of the readability of your build scripts. Consider the following script:

```
dependencies {
    implementation('org.springframework.boot:spring-boot-dependencies:1.5.8.RELEASE')
    implementation('org.springframework.boot:spring-boot-starter-web')
}
```

There's no indication that the dependency on `spring-boot-dependencies` will be treated any differently to the dependency on `spring-boot-starter-web`.

> Especially, it's not clear for everybody what a BOM is, because it's specified out of what is found in a POM file rather than on what it's supposed to represent

I would expect the concept of what a bom is to be clear to anyone who's trying to use one to provide dependency management in their build. They may not know or care that it's part of what's found in a pom file, they just know that it's something that's addressable via group ID, artifact ID, and version. A DSL that makes it clear that those coordinates are being used to influence a build's dependency management doesn't change that, but it does change how expressive the build script is.

> So I'd argue that if it's necessary to make the difference, we now have the tooling for this, also shipped in 4.6: for each dependency, you can add a "reason" why it's here:
>
> ```
> dependencies {
>    implementation('org.springframework.boot:spring-boot-dependencies:1.5.8.RELEASE') {
>       because "this application relies on the Spring Boot platform"
>    }
>}
>```

It's not the reason for why it's there that I found lacking. What was lacking for me is something that makes it clear that this dependency is special and is going to affect my build in ways that normal dependencies do not. The `// Import a bom` comment gave a hint that the following dependency was special. The reason does not. I would argue that if a DSL requires the use of a comment or because, then the DSL is not as expressive as it could be.

> In short, we want users to focus on why they need dependencies/BOMs, so that it's easier to understand when something is wrong.

I share that goal, which is why I think it's important that the DSL naturally makes it clear that a dependency is going to be used as a bom and affect a project's dependency management. This would then allow any comment or use of because to focus on the why and not the what. Put another way, it should be possible to move from one project built with Gradle to another and, with no prior knowledge of any of the project's dependencies, tell from the build script which dependencies are boms that will affect the project's dependency management.

**Comment by melix** [comment]
> What was lacking for me is something that makes it clear that this dependency is special and is going to affect my build in ways that normal dependencies do not. 

In practice they are not different. There would be no difference between "import bom" and adding a regular dependency on a BOM. I understand _why_ you want this, and we've also considered it as I said, but in practice it means that you have 2 syntaxes to do the same thing.

If you go that route, you could also argue that BOMs which have "optional" dependencies should be imported differently, because the optional dependencies _may_ influence the resolution of the graph depending on whether other direct or transitive dependencies on the same modules are found. This would be inconsistent, because what you add to the graph is a module, a platform declaration. You shouldn't have to look into the metadata file, to recognize it's a BOM or that it has optional dependencies to choose what syntax to use to consume them.

I do recognize that a dependency on a platform is somehow different from a dependency on a "hard module", because the only thing it does (in theory, it really depends on the BOMs you import in practice) is suggesting dependency versions. However, this is a _dependency on a platform_, so it is, effectively, a dependency. That's why, instead of having 2 syntax that would be exchangeable and have no effect whatsoever on the graph, we prefer to use the same syntax and add "reasons".

Note that reasons are not poor modeling, as you seem to imply. It's actually very important to get, as when you trace where a dependency comes from, be it using the dependency insight report or build scans, it will give you the trace of what happened, like:

- you depend on the Awesome Platform 1.0.0
- but version library 1.0.1 was rejected _because_ it has a CVE
- and it was upgraded to 1.0.2 _because_ of conflict resolution

and it also works with _rules_ that can explain _why_ they do something. A comment, or a DSL method, cannot provide that kind of information alone, as it's primarily aimed at humans.

**Comment by wilkinsona** [comment]
> In practice they are not different. There would be no difference between "import bom" and adding a regular dependency on a BOM.

This sounds like it's the crux of the matter. There should be a difference between importing a bom and adding a regular dependency on a bom. There's a difference in Maven, and therefore, everyone who authors boms expects those differences to be honoured when they're consumed.

Consider this pom which imports Spring Framework's bom:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<groupId>com.example</groupId>
	<artifactId>bom-import</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<packaging>jar</packaging>

	<dependencyManagement>
		<dependencies>
			<dependency>
				<groupId>org.springframework</groupId>
				<artifactId>spring-framework-bom</artifactId>
				<version>4.3.14.RELEASE</version>
				<type>pom</type>
				<scope>import</scope>
			</dependency>
		</dependencies>
	</dependencyManagement>

</project>
```

The relevant parts of the resulting effective pom are the following:

```xml
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.example</groupId>
  <artifactId>bom-import</artifactId>
  <version>0.0.1-SNAPSHOT</version>
  <dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-aop</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-aspects</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-beans</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-context</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-context-support</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-core</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-expression</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-instrument</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-instrument-tomcat</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-jdbc</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-jms</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-messaging</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-orm</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-oxm</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-test</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-tx</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-web</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-webmvc</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-webmvc-portlet</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
      <dependency>
        <groupId>org.springframework</groupId>
        <artifactId>spring-websocket</artifactId>
        <version>4.3.14.RELEASE</version>
      </dependency>
    </dependencies>
  </dependencyManagement>
  <repositories>
    …
  </repositories>
  <pluginRepositories>
    …
  </pluginRepositories>
  <build>
    …
  </build>
  <reporting>
    …
  </reporting>
</project>
```

Note that there are no `<dependencies>`. The imported bom has simply been inlined into the pom's `<dependencyManagement>`. Here's the module's dependency tree:

```
com.example:bom-import:jar:0.0.1-SNAPSHOT
```

By contrast, consider this pom which depends on Spring Framework's bom:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>

	<groupId>com.example</groupId>
	<artifactId>bom-dependency</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<packaging>jar</packaging>

	<dependencies>
		<dependency>
			<groupId>org.springframework</groupId>
			<artifactId>spring-framework-bom</artifactId>
			<version>4.3.14.RELEASE</version>
			<type>pom</type>
			<scope>import</scope>
		</dependency>
	</dependencies>

</project>
```

The resulting effective pom is the following:

```xml
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <groupId>com.example</groupId>
  <artifactId>bom-import</artifactId>
  <version>0.0.1-SNAPSHOT</version>
   <dependencies>
    <dependency>
      <groupId>org.springframework</groupId>
      <artifactId>spring-framework-bom</artifactId>
      <version>4.3.14.RELEASE</version>
      <type>pom</type>
      <scope>compile</scope>
    </dependency>
  </dependencies>
  <repositories>
    …
  </repositories>
  <pluginRepositories>
    …
  </pluginRepositories>
  <build>
    …
  </build>
  <reporting>
    …
  </reporting>
</project>
```

Note that there's no `<dependencyManagement>` and there's a single dependency upon a pom artifact. Here's the module's dependency tree:

```
com.example:bom-dependency:jar:0.0.1-SNAPSHOT
\- org.springframework:spring-framework-bom:pom:4.3.14.RELEASE:compile
```

> If you go that route, you could also argue that BOMs which have "optional" dependencies should be imported differently, because the optional dependencies may influence the resolution of the graph.

When you import a bom, any dependencies (optional or otherwise) declared in that bom are ignored. It's only the dependency management that is inlined.

> You shouldn't have to look into the metadata file, to recognize it's a BOM or that it has optional dependencies to choose what syntax to use to consume them.

As far as I can tell, this is exactly what 4.6 RC1 requires users to do.

There's no way in the build script to distinguish between a dependency that affects a project's dependency management and a dependency that affects a project's "normal" dependencies so I have to look at the metadata for the dependency and figure out what Gradle's going to do with it. Consider the following:

```
dependencies {
    implementation 'org.springframework.restdocs:spring-restdocs-core:2.0.0.RELEASE'
    implementation 'org.jruby:jruby-complete'
}
```

This fails as there's no version for `org.jruby:jruby-complete` despite the pom for `spring-restdocs-core` containing dependency management for `jruby-complete`. That actually makes sense to me as it aligns with Maven's behaviour for such a dependency, but it's completely unclear from the build script what the behaviour will be. To figure it out (assuming I have), you have to understand that:

1. Gradle treats poms with `<dependencyManagement>` and no `<dependencies>` differently to how it treats poms with `<dependencies>` or it treats poms with jar packaging differently to those with pom packaging.
2. `spring-restdocs-core` contains both `<dependencies>` and `<dependencyManagement>` or uses jar packaging, so its `<dependencyManagement>` is ignored.

> Note that reasons are not poor modeling, as you seem to imply. It's actually very important to get, as when you trace where a dependency comes from, be it using the dependency insight report or build scans, it will give you the trace of what happened

On the contrary, I think reasons look very useful. However,a user should be able to use them to describe why a dependency is there, rather than having to explain what the dependency is going to do (import some dependency management vs depend on the dependency and all of its transitive dependencies). A DSL which embraced the difference between importing and depending on a bom, and that allowed a reason to be specified when importing a bom would, I think, provide the best of both worlds.

**Comment by melix** [comment]
That's correct. There **is**, in Maven, a difference between the two, because BOMs are used for different purposes. This is why we typically introduced the "dependency constraints" (published, consumable) to solve that problem.

However, yes, there's a difference between a regular dependency, and a dependency on a platform. We're still thinking/modeling what it means in practice. Because we don't want it wrong, we're specifically not addressing that issue right now. It means that we _might_ come up with a syntax to "depend on a platform", like:

```
dependencies {
    usePlatform '...'
}
```

but that wouldn't be specific to Maven BOMs. So far all the use cases we have using dependency constraints can be solved without the need for a specific syntax. It doesn't mean that we're not going to introduce one, but if we do, it would cover both BOMs and "Gradle" platforms.

**Comment by wilkinsona** [comment]
FWIW, a syntax like that would address almost all of my concerns. I'm not looking for something that's bom-specific, just something that makes it clear that this is a different type of dependency and that Gradle will use it differently to a normal dependency.

>  So far all the use cases we have using dependency constraints can be solved without the need for a specific syntax.

At the risk of repeating myself, I am not arguing about whether or not the use cases can be solved, I'm arguing that the current syntax is ambiguous and makes build scripts harder to understand. I think it could be improved considerably by embracing the difference between depending on a regular dependency and depending on a platform.

I think I've now used (more than) enough of my time and yours on this issue. Thanks for the discussion. I think you now understand the reasoning behind the change that I'd like to see so I'll leave you in peace to make the right decision for Gradle as a whole.

**Comment by oehme** [comment]
> Gradle treats poms with <dependencyManagement> and no <dependencies> differently to how it treats poms with <dependencies> or it treats poms with jar packaging differently to those with pom packaging.

That's the crux and something we will change in 5.0 (correct me if our plans changed here @melix). We will treat all POMs the same. Every `<dependencyManagement>` will be treated as a constraint and every `<dependency>` will be treated as a hard dependency.

The reason we did special handling for BOMs was only to lower the potential performance impact of the many additional constraints added to the graph, but we can mitigate that.

**Comment by melix** [comment]
Hi all and thanks for the comments here,

## TL/DR

We’re going to add a way to “import a BOM”, using a nicer syntax. What it’s going to do is to explain to Gradle that the _variant_ we’re resolving is the platform variant, not the library variant, when a POM is used to represent the 2 things. Something around those lines:

```
dependencies {
   api platform('org:platform:1.0')
}
```

## The problem

In order to understand properly what we will decide, I first need to expose the differences between Maven’s `<dependencyManagement/>` block and Gradle’s `constraints`. They may look similar at first glance, but there are key differences:

* both Maven’s dependency management block and Gradle’s dependency constraints act as suggestions or constraints on versions of components _as they are found in the graph_. 
Maven’s dependency management block is taken in account:
    * for direct dependencies (dependency management blocks of transitive dependencies are only used to constraint the dependencies of that module)
    * in a parent POM
* Gradle’s dependency constraints are transitive (constraints of transitive dependencies are used during dependency resolution)

To do best effort with regards to Maven compatibility, when parsing POMs, Gradle, as in 4.6, makes some decisions:

* when we see a POM file which packaging is pom, then we add all dependency management dependencies as constraints
* dependencies of a POM file are always added, independently of the packaging

That means we consider something a BOM if, and only if, its packaging is `pom`. Now, the Gradle “dependency model” is not the same as Maven. There’s no such thing as an “import” scope, because it has no meaning in Gradle. We have “configurations”, which correspond to usages. For example, adding an “api” dependency, or an “implementation” dependency. A dependency brings in a module, which can be a library, or, what we call a “platform”. It means that if you do:

```
dependencies {
    api ‘some:platform:1.0’
}
```

The semantics for Gradle are pretty clear: the component at some:platform has a version, 1.0, and it happens to be a platform, which can bring in additional constraints (we’ll assume for now that platform is published with Gradle). Those constraints would only apply on the “api” configuration, because they only make sense in that context. This means that if you have functional tests, or a binary compatibility check configuration, the use of the constraints wouldn’t leak into those configurations. This point is to address why we don’t have a single “import” notation that applies on all configurations. We think it’s not correct to model it that way, and there’s an easy workaround:

```
// first workaround, be explicit, which is probably better
dependencies {
    api ‘some:platform:1.0’
    implementation ‘some:platform:1.0’
}

// second workaround, add the dependency to all configurations that make sense:

dependencies {
    configurations.all { conf ->
       // this is not quite right because we’re adding a dependency on configurations which do not matter
       add(conf, ‘some:platform:1.0’)
    }
}
```

The second point to understand, and that directly impacts our design decisions, is _why_ dependency management blocks or constraints are used. They basically fall into 2 categories:

* for a library, as an “implementation detail”. This is what most Maven users are used to: a `<dependencyManagement>` block defines a bunch of library versions, which are shared by different modules. Then, when published, the only thing that _should_ matter is the resolved versions. In the Maven world, those dependency management blocks are used when a module is built (and they influence transitive dependencies), but not when it’s consumed.
* for a platform, to “suggest dependency versions”. The typical use case is the Spring Boot platform BOM.

However, in the Maven world, you don’t explicitly state which of the 2 categories you are. This means that you can have a library, published with `<packaging>jar</packaging>`, which effectively has a `<dependencyManagement>` block. The design decisions made in 4.6 implied that this was an “implementation detail”. However, there are some people who also rely on this component _as a platform_. Said differently, there are two different components living at the same GAV coordinates:

* the library
* the platform

And when in Gradle, you consume this:

```
dependencies {
    api ‘some:library:1.0’
}
```

It will assume that you want the library, not the platform. So dependency management block is ignored. We ignore it for a good reason: as explained, it’s most often the case that the dependency management block here was an internal implementation detail of building the library, that shouldn’t leak to the consumers. The constraints do _not_ have to be applied transitively, and that’s the reason why we don’t systematically convert all the dependency management blocks into constraints. In the Gradle world, you would model this differently, by having 2 different components: one that is the platform, and the other that is the library. Both would live at different coordinates, and when you _depend on_ one of them, you know what you bring in.

So this leaves us with a problem: when the component is published by Maven **and** that it happily mixes the platform and the library in the same GAV, we are potentially wrong, which is exactly why this ticket was created.

So here are the potential solutions to this problem, which basically boils down to how to select the appropriate component (platform or library) from something published at a single GAV. If you’re only interested in what we’re likely to implement, go with solution #6.

### Solution 1: Do nothing

Consider that our current implementation is the right trade-off. In other words, if a Maven platform wasn’t published using `pom` packaging on its own coordinates, we cannot use it. Obviously, this solution is not satisfying as we have several examples in this issue of people using libraries as BOMs.

### Solution 2: Consider all dependency management blocks as constraints

Independently of the packaging, whenever we see a dependency management block, consider it as if it was a constraint. That was our initial implementation, but we reverted it for several reasons:
as explained above, this is a wrong model of things, because in most cases, the dependency management block is just an implementation detail of the publisher, and it shouldn’t have been published in the first place. Because Maven uses the same POM for producing and for consumers means that when you consume, you don’t know what was the intent.
it can significantly change the result of dependency resolution. Because in the Gradle world, constraints are transitive, it means that dependency management blocks of transitive dependencies would be considered constraints, and influence resolution. This has a significant risk of breaking changes, as well as behavioral differences between Maven and Gradle.
it has an impact on performance. Because <dependencyManagement> blocks are pretty frequent in Maven, and because they are used as implementation details more than actual BOMs, they lead to more conflict resolution, which slows down things. First measurements led up to 25% slowdown on a very large, real world dependency graph

### Solution 3: Add Maven-specific disambiguation

The idea is to do what was initially suggested in this ticket. Because we have an ambiguity, we need to fix it. A good way to fix it is to be explicit, and tell Gradle “hey, I’m using a BOM, do the right thing”. So basically, when seeing a specific notation as a dependency, Gradle would use the target component _as a_ BOM. It means it would get its dependency management block and apply constraints (bonus point: should it also apply the dependencies if it’s a library-as-a-bom?).

```
dependencies {
     // syntax proposal 1
     importBom(‘api’, ‘some:platform:1.0)
     // syntax proposal 2
     api bom(‘some:platform:1.0’)
}
```

While it solves the problem, this approach has several consequences:

* all transitive dependencies will ignore dependency management blocks, which is, probably what you would expect
* it cannot be easily published in Gradle metadata: this is a Maven specific workaround. We don’t have any special kind of dependency, all we have is “dependencies as components”. So when published, this would (probably) become a regular dependency, which would then be ignored by consumers.
since it’s on the consumer side, it means we need to access the original metadata, and add to the graph different elements based on what is declared on the consumer side, which is high complexity, in particular in terms of cached metadata and how to interact with the dependency graph builder.

The other problem is that this is a very Maven-specific notation, to fix a Maven model deficiency. This can be mitigated by only adding the syntax if we actually apply a “maven-compatibility” plugin, or something like that.

### Solution 4: special dependency notation for BOMs

This is really a variant of 3. Just like we have classifiers, we _could_ use the dependency notation to split heads, and create two components off this dual “library/BOM” POM. In other words, we would create two different components from the same GAV. Something like this:

```
dependencies {
    api ‘some:library:1.0@@as-bom’
}
```

Then, we would internally have 2 components. One that represents the library, where the dependency management block would be ignored (as constraints). And the other, which would be the BOM part, containing only the constraints. It solves some implementation problems because we don’t need to make the difference between optional dependencies and constraints anymore, since we would have 2 real components. However there are still drawbacks:

technically this forces us to bake in rules that there’s no longer a single component living at some GAV. This has serious technical challenges which go against the simpler solution of using variant-aware dependency management (see solution 6)
when published, what do we use? Obviously this is not a valid dependency syntax, especially for Maven consumers. Should we ignore the BOM part and “loose information”?

### Solution 5: options for including/excluding dependencies/constraints

With this, we would add a way, for a dependency declaration, to explain whether we want to consider the constraints, dependencies, or both:

```
dependencies {
    api (‘some:library:1.0’) {
        includeConstraints = true
        includeDependencies = true
    }
}
```

The problem now becomes a “defaults” problem:

* for Gradle metadata, the default be “true” for both constraints and dependencies, because we know that constraints need to be transitive (they are not implementation details)
* for Maven, the default would be true for dependencies and false for constraints. Then, we would parse metadata, include everything we know about the component, and defer the decision to include constraints based on those flags. This, again, implies making the difference between an optional dependency and a constraint (our internal sauce).
can users deal with the complexity of knowing why the default are different based on what metadata we got? (Maven vs Gradle)
this would also imply publishing those flags in Gradle metadata, so that we honor them for consumers

The advantage, on the other hand, is that it can solve _bad Gradle metadata too_. In other words, someone publishing constraints when all they wanted is dependencies.

### Solution 6: Use variant-aware dependency management

Gradle has variant-aware dependency management. In this world, a component can have several variants. For example, a component may expose an API, for compile, and a runtime, for execution. Variants have different dependencies, and the engine selects between the variants based on the requirements. This is already working internally, and used widely in the Android ecosystem.

When we publish a Java library (using the `java-library`) plugin, 2 variants of the library are found in module metadata: the “api” variant and the “runtime” variant. Similarly, when we see a Maven POM, we actually map it to variants, one named “compile”, the other named “runtime”, which are also used for API and implementation.

Now, the idea here would be that when we parse Maven metadata, we create an additional variant, the “platform” variant. This variant would have constraints based on the `<dependencyManagement>` block. The “compile” and “runtime” variants would be untouched. 

Then now it’s just a matter of consuming the right variant. When a dependency is declared, you would add an attribute to express the requirement on the platform (BOM) rather than the library:

```
dependencies {
    api (‘org:library:1.0) {
        attributes {
           attribute(PLATFORM_ATTRIBUTE, true)
        }
    }
}
```

This would solve the problem without much hassle, but it also has several consequences:

* the syntax is not very user-friendly, but we could imagine a shortcut notation to add the attribute transparently:
   * api platform(‘...’)
* attributes on dependencies are not supported yet (this is WIP, though): you can only add attributes to _configurations_ right now. So we cannot implement this solution right now, but it shouldn’t be too long before we can.

Interestingly, it also maps nicely with Gradle concepts. As we explained, when we publish a Java library, _today_ what we do is publishing the constraints as part of the library variants (`api`, `implementation`). However, we _could_ decide not to publish them, and only do it when the component which is published is actually a platform (`apply plugin: ‘platform’`), *or* using a Java library with explicit opt-in to publish the constraints (either using: `from components.javaLibraryPlatform` or tweaking the `javaLibrary` component).

At this point we are quite convinced the best solution is solution 6.

**Comment by wilkinsona** [comment]
It's clear that you've given this a lot of thought and careful consideration. Thank you. FWIW, I like solution 6, particularly if the shortcut notation is implemented. The separate configuration of the platform attribute is, by Gradle's standards, unusually verbose.

**Comment by oehme** [comment]
@melix is this still planned for 4.8?

**Comment by ljacomet** [comment]
No, it has been pushed back. I have cleared the milestone.

**Comment by melix** [comment]
Hi @wilkinsona ,

Finally, the [PR for this is ready for review](https://github.com/gradle/gradle/pull/6425). It implements solution 6 described above. If you have time, I'd appreciate if you could checkout the branch and give it a try. Everything is still hidden behind the experimental flag, but here's what happens now:

1. if we add a regular dependency on a BOM (that is to say, a Maven POM with packaging `pom`, a `dependencyManagement` block and no regular dependencies, basically nothing will happen, as we would add a dependency on a component which has no dependencies. eg:

```
dependencies {
   implementation "my:bom:1.0" // declares a dependency on a BOM, but it's a regular dependency, nothing will happen
}
```

2. if, instead, we use the `platform` keyword in the dependency declaration, the `dependencyManagement` block of the POM will be used to define _dependency constraints_ in Gradle:

```
dependencies {
   implementation platform("my:bom:1.0") // declares a dependency on the platform, the dependency management section of the target POM will be used as a source for constraints
}
```

3. If a component published at some GAV is not a BOM, but contains both a dependency management section and a dependencies section, then you can choose between the "platform" component or the "library" component by using the `platform` keyword:

```
dependencies {
   implementation "some:component:1.0" // will use the dependencies section of the POM as a source for constraints, but NOT the dependency management block, which is now going to be ignored completely
   // or...
   implementation platform("some:component:1.0") // will use the dependencyManagement section of the POM as a source for constraints
}
```

4. Last but not least, we also added the concept of "enforced platform". If you use the `enforcedPlatform` keyword instead of `platform`, the versions found in the dependency management block are effectively forced, overriding any other version found in the graph.

```
dependencies {
   implementation enforcedPlatform("my:bom:1.0") // declares a dependency on the platform, but the versions found in the BOM will override any other version found in the graph
}
```

**Comment by wilkinsona** [comment]
@melix Thank you for asking me to take a look and thanks for your patience while I was digging myself out of a post-vacation backlog. From your description, the changes sound like a big step forwards. I see they've now been merged. I'll try them out properly some time this week.

**Comment by melix** [comment]
@wilkinsona Just a heads-up: `master` now has improved POM support enabled by default. This means that in practice you won't need the experimental flag enabled for this to work using the latest nightly (5.0-20180919145027+0000).

**Comment by melix** [comment]
@wilkinsona Sorry to bother you during S1P, did you have a chance to test this? We're waiting for your feedback to close this issue, and 5.0 is getting closer. If I don't hear anything by the end of the week I'll optimistically close as resolved.

**Comment by melix** [comment]
Thanks for the feedback Andy. That you get a nesting level with the native platform support is expected. With Gradle, platforms are a "thing", so it makes sense (in particular if you consider alignment into the game).

With regards to the dependencies report, I get that it can be confusing. Note that we will _not_ work on improving this report, as it's superseded by build scans. Here's for example your build script output in a build scan: https://scans.gradle.com/s/kc4eqdidruwy6/dependencies?focusedDependency=WzAsMCwxNjQsWzAsMCxbMTY0XV1d&toggled=W1swXSxbMCwwXV0

On the CLI side of things it's better to use the `dependencyInsight` report task, which would give you more information: https://scans.gradle.com/s/qqyc44zjkkio4/console-log

Regarding ordering I'm not sure we can do anything right now. This is a consequence of the resolution engine algorithm itself, but I'll take a look.

**Comment by ljacomet** [comment]
After looking into finding a quick fix, there is no easy solution to that classpath ordering problem. Given that the comments are overall positive on the other aspects, closing this.
